### PR TITLE
sleep instead of drop when stream rate exceeded limit;

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
Currently we are dropping the stream right away when the stream rate limit is being exceeded. The problem is the client can retry right away. In addition, the stream might have been received and can be used, dropping it waste resources.

#### Summary of Changes

Change to sleep instead of drop. Thanks to @alessandrod -- we sleep based on the estimated time to elapse to appease the rate limit and consider the RTT for the connection.

Fixes #

### CODE REVIEW

File: streamer/src/nonblocking/stream_throttle.rs
Comment (alessandrod) [inline streamer/src/nonblocking/stream_throttle.rs:219]: It was already the case before this PR so doesn't block this PR, but this is
racy: we're doing a check with a read lock, then dropping and acquiring a write
lock. Multiple threads can find that they're > THROTTLING_INTERVAL with the read
lock and update the last throttling instant.

I think that this connection-wide limit can be enforced without locks. Every connection task keeps
its last_throttling_instant and stream_count. The max streams allowed per connection is scaled by
the number of connections opened by a peer, which can be passed as an atomic to each connection
task.

**Review by alessandrod — COMMENTED** [review]
lgtm, but someone else should review since I wrote some of the original code

File: streamer/src/nonblocking/stream_throttle.rs
Comment (lijunwangs) [reply streamer/src/nonblocking/stream_throttle.rs:219]: I created a separate issue to address as it is orthogonal to this PR.

Comment (lijunwangs) [reply streamer/src/nonblocking/stream_throttle.rs:219]: https://github.com/anza-xyz/agave/issues/942

Comment (pgarg66) [reply streamer/src/nonblocking/stream_throttle.rs:219]: It should not be racy, as we check again for the timeout after acquiring the write lock. Only one thread will acquire the write lock at a time, and the first one to acquire will update the `last_throttling_instant`.

File: streamer/src/nonblocking/quic.rs
Comment (pgarg66) [inline streamer/src/nonblocking/quic.rs]: Can we explain in a comment why we are subtracting `connectio.rtt()`? It will help with future maintainability of the code.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: The comment could say something like:

```
// ... so the peer backs off.
//
// We subtract `connection.rtt()` to the time we sleep to minimize drift. That's the estimated amount
// of time needed for us to send credits back to the peer, and for the peer to send the next chunk of
// data so that it falls near the start of the next interval window.
```

Comment (pgarg66) [reply streamer/src/nonblocking/quic.rs:837]: If that's the reason, we should reduce the throttle duration just before calling sleep. The current code is actually increasing the total throttling window by the rtt time

Comment (pgarg66) [reply streamer/src/nonblocking/quic.rs]: What's average value for rtt. Wondering if this will reduce the throttling window enough that throttling never happens.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: It's per connection and the quic protocol describes how to compute and update it. It's meant to be a good estimate of the actual rtt and is used internally by quinn to piggyback control frames to minimize overhead. It can only be way off if there's a bug in quinn, but at that point, many other things wouldn't work as intended anyway.

Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: Thinking more about this, I think the sleep value should be the amount of interval which will make this packet itself to satisfy the rate limit. And I do not think we need to think about the  RTT which is for the next packet. And the value of the sleep of STREAM_THROTTLING_INTERVAL - the elapse of the interval is also questionable. Maybe that is too large actually. From the math we should be able to figure out the amount of the time to sleep to satisfy the rate limit requirement.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: We want to read X streams each interval of T ms. We just found that we've read X streams and we're at time T1<T.

Ideally, we wanted to have received this stream X at time T, not T1. So we're going to sleep to slow down the peer, and we want them to send the next stream in the *next* interval window of T ms. The difference between the time we received the stream and the time we wanted to receive it is T-T1. That's STREAM_THROTTLING_INTERVAL -  throttle_interval_start.elapsed().

If we sleep for T - T1, *we* will send a frame back to the peer at time T which tells them that they can start sending again. That will take rtt / 2 time to reach the peer, which will start sending again, so we'll receive the next stream around T + rtt. We want to receive the next stream *approximately* at time T tho (the beginning of the next interval), therefore we also subtract rtt.

Where is the logic wrong?

Comment (pgarg66) [reply streamer/src/nonblocking/quic.rs]: For a well behaving client it would be ok. I am wondering if a malicious client can implement/hack a custom quic client that causes high rtt but also has high stream count. A simple mitigation would be to cap the rtt value used in this calculation.

Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: The consideration of RTT makes sense only when you assume the streams are sent sequentially. It is not true, the client can send the streams simultaneously. While we think it will take a RTT to get the next stream, the next stream might have already arrived. I think the logic is simpler just consider this stream which is causing the violation -- not do the optimization for the next stream which we do not know when it might come.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: > For a well behaving client it would be ok. I am wondering if a malicious client can implement/hack a custom quic client that causes high rtt but also has high stream count. A simple mitigation would be to cap the rtt value used in this calculation.

I'll check the spec, but yeah you're right in principle a malicious client could intentionally game the rtt to make his connection more bursty. Not sure what a good cap for rtt could be.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: @lijunwangs 
>  While we think it will take a RTT to get the next stream, the next stream might have already arrived

No, because by sleeping if the peer keeps sending the receive window will fill up, which is the whole point of sleeping. So the next stream _the peer sends_ will be after our sleeping time. If in the meantime there are queued streams within the receive window that's fine.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: > Not sure what a good cap for rtt could be.

I say for the purpose of landing this soon, let's remove rtt as a variable. Then we can play with taking rtt in consideration for computing the receive window and/or the throttle time and we can let that ride the trains.

Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: done

**Comment by pgarg66** [comment]
LGTM. It has a merge conflict right now.

**Comment by lijunwangs** [comment]
> LGTM. It has a merge conflict right now.

I have rebased the master and addressed the conflicts. @pgarg66

**Review by pgarg66 — APPROVED** [review]

**Comment by t-nelson** [comment]
no blockers here, right?

**Pull request merged**
